utils/memory_bench/src/main.rs

```diff
@@ -0,0 +1,421 @@
+        panic!(
+            "Unable to create the benchmark directory {:?} {:?}",
+            benchmark_dir, err
+        );
+    });
+
+    if benchmark_output_path.exists() {
+        fs::remove_file(&benchmark_output_path).unwrap_or_else(|err| {
+            panic!(
+                "Could not remove the file: {:?} {:?}",
+                benchmark_output_path, err
+            );
+        });
+    }
+
+    let mut benchmark_output = fs::OpenOptions::new()
+        .read(true)
+        .append(true)
+        .create(true)
+        .open(&benchmark_output_path)
+        .expect("Unable to open the benchmark output file for write.");
+
+    for example in examples {
+        println!("[memory_bench] Starting example {:?}", example);
+        let example_path = {
+            let mut path = example_dir.clone();
+            path.push(format!("{}.rs", example));
+            if !path.exists() {
+                eprintln!("Could not find the example at the path: {:?}", path);
+                process::exit(1);
+            }
+            path
+        };
+        let example_copy_path = {
+            let mut path = example_dir.clone();
+            path.push(format!("{}___memory.rs", example));
+            path
+        };
+
+        println!(
+            "[memory_bench] Injecting the memory instrumentation into {:?}",
+            example
+        );
+        let instrumented_src = run_dhat_injection(&example_path);
+
+        let mut file = File::create(&example_copy_path).unwrap_or_else(|err| {
+            panic!(
+                "Could not create the file {:?} {:?}",
+                example_copy_path, err
+            );
+        });
+
+        file.write_all(instrumented_src.as_bytes())
+            .unwrap_or_else(|err| {
+                eprintln!("Could not write the file {:?} {:?}", example_copy_path, err);
+                process::exit(1);
+            });
+
+        println!(
+            "[memory_bench] Running the instrumented example {:?}",
+            example
+        );
+
+        let mut run_example = Command::new("cargo")
+            .arg("+nightly")
+            .arg("run")
+            .arg("--example")
+            .arg(format!("{}___memory", example))
+            .arg("--profile")
+            .arg("bench")
+            .arg("-Z")
+            .arg("unstable-options")
+            .current_dir(&component_dir)
+            .stderr(Stdio::piped())
+            .spawn()
+            .unwrap_or_else(|err| {
+                eprintln!("The example {:?} failed to run. {:?}", example, err);
+                process::exit(1);
+            });
+
+        let stdout = run_example
+            .stderr
+            .take()
+            .expect("No stderr in the example.");
+
+        let dhat_log: Vec<_> = BufReader::new(stdout)
+            .lines()
+            .map(|s| s.expect("Unable to read from stderr."))
+            .inspect(|s| println!("[memory_bench] > {}", s))
+            .filter(|s| s.starts_with("dhat: "))
```

Comment:
Can you interact with dhat more programmatically?  It seems "hacky" to read the results from stderr.

Reply:
I believe the `dhat-heap.json` file could contain this information, if you know how to parse it and read it. It contains an allocation tree summarizing the allocations. The format is a bit dense, but I wrote an importer for the Firefox Profiler trying to understand this data format. It looks like the data might be available if I parsed the JSON, and iterated over it.

Here is the relevant type definition I created for the format in my importer: https://github.com/firefox-devtools/profiler/pull/3128/files#diff-9decaaf66238dad5abffb33cb8ab16232709f9a7b6ce911084489763834aac56R92-R138

Maybe this could be a follow-up? It would require a bit of examination to see if I could pull out the same numbers.

Reply:
My plan is to follow-up with this one, rather than handle it in the initial PR. https://github.com/unicode-org/icu4x/issues/460